aiohttp/web_request.py

```diff
@@ -690,6 +694,21 @@ def __bool__(self) -> bool:
 
     def _cancel(self, exc: BaseException) -> None:
         self._payload.set_exception(exc)
+        for fut in self._disconnection_waiters:
+            fut.set_result(None)
```

Comment:
Welcome to low-level asyncio!
This line contains a very common but annoying bug. Let me describe the failure scenario:
1. a task calls `await request.wait_for_disconnection()` and waits
2. the task gets canceled by explicit `task.cancel()` call
3. eventually the connection is broken by peer. `request._cancel()` is called; which in turn calls `fut.set_result(None)`.
4. The future is canceled by bullet 2, `set_result()` fails.

Please use `helpers.set_result()`.

Reply:
Ah, good catch.  Changed to use helpers.set_result.

Reply:
versionadded...
Note: this PR needs rebase/squash commits before merging.